File: packages/authgear-capacitor/android/src/main/java/com/authgear/capacitor/AuthgearPlugin.java
package com.authgear.capacitor;

import android.app.Activity;
import android.content.Context;
import android.content.Intent;
import android.net.Uri;

import androidx.activity.result.ActivityResult;
import androidx.annotation.NonNull;
import androidx.appcompat.app.AppCompatActivity;
import androidx.biometric.BiometricManager;
import androidx.biometric.BiometricPrompt;

import com.getcapacitor.JSArray;
import com.getcapacitor.JSObject;
import com.getcapacitor.JSValue;
import com.getcapacitor.Plugin;
import com.getcapacitor.PluginCall;
import com.getcapacitor.PluginMethod;
import com.getcapacitor.annotation.ActivityCallback;
import com.getcapacitor.annotation.CapacitorPlugin;

import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

@CapacitorPlugin(name = "Authgear")
public class AuthgearPlugin extends Plugin {

    private Authgear implementation = new Authgear();

    @PluginMethod
    public void storageGetItem(PluginCall call) {
        String key = call.getString("key");
        try {
            Context ctx = this.getContext();
            String value = this.implementation.storageGetItem(ctx, key);
            JSObject ret = new JSObject();
            if (value == null) {
                ret.put("value", JSONObject.NULL);
            } else {
                ret.put("value", value);
            }
            call.resolve(ret);
        } catch (Exception e) {
            this.reject(call, e);
        }
    }

    @PluginMethod
    public void storageSetItem(PluginCall call) {
        String key = call.getString("key");
        String value = call.getString("value");
        try {
            Context ctx = this.getContext();
            this.implementation.storageSetItem(ctx, key, value);
            call.resolve();
        } catch (Exception e) {
            this.reject(call, e);
        }
    }

    @PluginMethod
    public void storageDeleteItem(PluginCall call) {
        String key = call.getString("key");
        try {
            Context ctx = this.getContext();
            this.implementation.storageDeleteItem(ctx, key);
            call.resolve();
        } catch (Exception e) {
            this.reject(call, e);
        }
    }

    @PluginMethod
    public void randomBytes(PluginCall call) {
        Integer length = call.getInt("length");
        try {
            byte[] bytes = this.implementation.randomBytes(length.intValue());
            JSArray array = new JSArray();
            for (byte b : bytes) {
                array.put(b);
            }
            JSObject ret = new JSObject();
            ret.put("bytes", array);
            call.resolve(ret);
        } catch (Exception e) {
            this.reject(call, e);
        }
    }

    @PluginMethod
    public void sha256String(PluginCall call) {
        String input = call.getString("input");
        try {
            byte[] bytes = this.implementation.sha256String(input);
            JSArray array = new JSArray();
            for (byte b : bytes) {
                array.put(b);
            }
            JSObject ret = new JSObject();
            ret.put("bytes", array);
            call.resolve(ret);
        } catch (Exception e) {
            this.reject(call, e);
        }
    }

    @PluginMethod
    public void getDeviceInfo(PluginCall call) {
        try {
            Context ctx = this.getContext();
            JSONObject deviceInfo = this.implementation.getDeviceInfo(ctx);
            JSObject ret = new JSObject();
            ret.put("deviceInfo", deviceInfo);
            call.resolve(ret);
        } catch (Exception e) {
            this.reject(call, e);
        }
    }

    @PluginMethod
    public void generateUUID(PluginCall call) {
        String uuid = this.implementation.generateUUID();
        JSObject ret = new JSObject();
        ret.put("uuid", uuid);
        call.resolve(ret);
    }

    @PluginMethod
    public void openAuthorizeURL(PluginCall call) {
        String urlString = call.getString("url");
        Uri uri = Uri.parse(urlString).normalizeScheme();
        String callbackURLString = call.getString("callbackURL");
        Uri callbackURL = Uri.parse(callbackURLString).normalizeScheme();

        Context ctx = this.getContext();
        OAuthRedirectActivity.registerCallbackURL(callbackURL.toString());
        Intent intent = OAuthCoordinatorActivity.createAuthorizationIntent(ctx, uri);
        this.startActivityForResult(call, intent, "handleOpenAuthorizeURL");
    }

    @ActivityCallback
    private void handleOpenAuthorizeURL(PluginCall call, ActivityResult activityResult) {
        int resultCode = activityResult.getResultCode();
        if (resultCode == Activity.RESULT_CANCELED) {
            this.rejectWithCancel(call);
        }
        if (resultCode == Activity.RESULT_OK) {
            String redirectURI = activityResult.getData().getData().toString();
            JSObject ret = new JSObject();
            ret.put("redirectURI", redirectURI);
            call.resolve(ret);
        }
    }

    @PluginMethod
    public void openURL(PluginCall call) {
        String urlString = call.getString("url");
        Uri uri = Uri.parse(urlString).normalizeScheme();

        Context ctx = this.getContext();
        Intent intent = WebViewActivity.createIntent(ctx, uri.toString());
        this.startActivityForResult(call, intent, "handleOpenURL");
    }

    @ActivityCallback
    private void handleOpenURL(PluginCall call, ActivityResult activityResult) {
        int resultCode = activityResult.getResultCode();
        if (resultCode == Activity.RESULT_CANCELED) {
            call.resolve();
        }
        if (resultCode == Activity.RESULT_OK) {
            call.resolve();
        }
    }

    @PluginMethod
    public void checkBiometricSupported(PluginCall call) {
        JSObject android = call.getObject("android");
        JSONArray constraint = this.jsObjectGetArray(android, "constraint");
        int flags = this.constraintToFlag(constraint);

        Context ctx = this.getContext();
        try {
            int result = this.implementation.checkBiometricSupported(ctx, flags);
            if (result == BiometricManager.BIOMETRIC_SUCCESS) {
                call.resolve();
            } else {
                String resultString = this.resultToString(result);
                call.reject(resultString, resultString);
            }
        } catch (Exception e) {
            this.reject(call, e);
        }
    }

    @PluginMethod
    public void createBiometricPrivateKey(PluginCall call) {
        AppCompatActivity activity = this.getActivity();

        JSObject payload = call.getObject("payload");
        String kid = call.getString("kid");
        String alias = "com.authgear.keys.biometric." + kid;
        JSObject android = call.getObject("android");
        JSONArray constraint = this.jsObjectGetArray(android, "constraint");
        boolean invalidatedByBiometricEnrollment = android.getBool("invalidatedByBiometricEnrollment");
        int flags = this.constraintToFlag(constraint);
        String title = android.getString("title");
        String subtitle = android.getString("subtitle");
        String description = android.getString("description");
        String negativeButtonText = android.getString("negativeButtonText");

        BiometricOptions options = new BiometricOptions();
        options.payload = payload;
        options.kid = kid;
        options.alias = alias;
        options.flags = flags;
        options.invalidatedByBiometricEnrollment = invalidatedByBiometricEnrollment;
        options.title = title;
        options.subtitle = subtitle;
        options.description = description;
        options.negativeButtonText = negativeButtonText;

        try {
            this.implementation.createBiometricPrivateKey(
                    activity,
                    options,
                    new BiometricCallback() {
                        @Override
                        public void onSuccess(String jwt) {
                            JSObject obj = new JSObject();
                            obj.put("jwt", jwt);
                            call.resolve(obj);
                        }

                        @Override
                        public void onAuthenticationError(int errorCode, @NonNull CharSequence errString) {
                            call.reject(errString.toString(), AuthgearPlugin.this.errorCodeToString(errorCode));
                        }

                        @Override
                        public void onException(Exception e) {
                            AuthgearPlugin.this.reject(call, e);
                        }
                    }
            );
        } catch (Exception e) {
            this.reject(call, e);
        }
    }

    @PluginMethod
    public void signWithBiometricPrivateKey(PluginCall call) {
        AppCompatActivity activity = this.getActivity();

        JSObject payload = call.getObject("payload");
        String kid = call.getString("kid");
        String alias = "com.authgear.keys.biometric." + kid;
        JSObject android = call.getObject("android");
        JSONArray constraint = this.jsObjectGetArray(android, "constraint");
        boolean invalidatedByBiometricEnrollment = android.getBool("invalidatedByBiometricEnrollment");
        int flags = this.constraintToFlag(constraint);
        String title = android.getString("title");
        String subtitle = android.getString("subtitle");
        String description = android.getString("description");
        String negativeButtonText = android.getString("negativeButtonText");

        BiometricOptions options = new BiometricOptions();
        options.payload = payload;
        options.kid = kid;
        options.alias = alias;
        options.flags = flags;
        options.invalidatedByBiometricEnrollment = invalidatedByBiometricEnrollment;
        options.title = title;
        options.subtitle = subtitle;
        options.description = description;
        options.negativeButtonText = negativeButtonText;

        try {
            this.implementation.signWithBiometricPrivateKey(
                    activity,
                    options,
                    new BiometricCallback() {
                        @Override
                        public void onSuccess(String jwt) {
                            JSObject obj = new JSObject();
                            obj.put("jwt", jwt);
                            call.resolve(obj);
                        }

                        @Override
                        public void onAuthenticationError(int errorCode, @NonNull CharSequence errString) {
                            call.reject(errString.toString(), AuthgearPlugin.this.errorCodeToString(errorCode));
                        }

                        @Override
                        public void onException(Exception e) {
                            AuthgearPlugin.this.reject(call, e);
                        }
                    }
            );
        } catch (Exception e) {
            this.reject(call, e);
        }
    }

    @PluginMethod
    public void removeBiometricPrivateKey(PluginCall call) {
        String kid = call.getString("kid");
        String alias = "com.authgear.keys.biometric." + kid;

        try {
            this.implementation.removeBiometricPrivateKey(alias);
            call.resolve();
        } catch (Exception e) {
            this.reject(call, e);
        }
    }

    private JSONArray jsObjectGetArray(JSObject obj, String key) {
        try {
            return obj.getJSONArray(key);
        } catch (JSONException e) {
            throw new RuntimeException(e);
        }
    }

    private int constraintToFlag(JSONArray constraint) {
        try {
            int flag = 0;
            for (int i = 0; i < constraint.length(); i++) {
                String c = constraint.getString(i);
                if (c.equals("BIOMETRIC_STRONG")) {
                    flag |= BiometricManager.Authenticators.BIOMETRIC_STRONG;
                }
                if (c.equals("DEVICE_CREDENTIAL")) {
                    flag |= BiometricManager.Authenticators.DEVICE_CREDENTIAL;
                }
            }
            return flag;
        } catch (JSONException e) {
            throw new RuntimeException(e);
        }
    }

    private String resultToString(int result) {
        switch (result) {
            case BiometricManager.BIOMETRIC_ERROR_HW_UNAVAILABLE:
                return "BIOMETRIC_ERROR_HW_UNAVAILABLE";
            case BiometricManager.BIOMETRIC_ERROR_NONE_ENROLLED:
                return "BIOMETRIC_ERROR_NONE_ENROLLED";
            case BiometricManager.BIOMETRIC_ERROR_NO_HARDWARE:
                return "BIOMETRIC_ERROR_NO_HARDWARE";
            case BiometricManager.BIOMETRIC_ERROR_SECURITY_UPDATE_REQUIRED:
                return "BIOMETRIC_ERROR_SECURITY_UPDATE_REQUIRED";
            case BiometricManager.BIOMETRIC_ERROR_UNSUPPORTED:
                return "BIOMETRIC_ERROR_UNSUPPORTED";
            case BiometricManager.BIOMETRIC_STATUS_UNKNOWN:
                return "BIOMETRIC_STATUS_UNKNOWN";
            default:
                return "BIOMETRIC_ERROR_UNKNOWN";
        }
    }

    private String errorCodeToString(int errorCode) {
        switch (errorCode) {
            case BiometricPrompt.ERROR_CANCELED:
                return "ERROR_CANCELED";
            case BiometricPrompt.ERROR_HW_NOT_PRESENT:
                return "ERROR_HW_NOT_PRESENT";
            case BiometricPrompt.ERROR_HW_UNAVAILABLE:
                return "ERROR_HW_UNAVAILABLE";
            case BiometricPrompt.ERROR_LOCKOUT:
                return "ERROR_LOCKOUT";
            case BiometricPrompt.ERROR_LOCKOUT_PERMANENT:
                return "ERROR_LOCKOUT_PERMANENT";
            case BiometricPrompt.ERROR_NEGATIVE_BUTTON:
                return "ERROR_NEGATIVE_BUTTON";
            case BiometricPrompt.ERROR_NO_BIOMETRICS:
                return "ERROR_NO_BIOMETRICS";
            case BiometricPrompt.ERROR_NO_DEVICE_CREDENTIAL:
                return "ERROR_NO_DEVICE_CREDENTIAL";
            case BiometricPrompt.ERROR_NO_SPACE:
                return "ERROR_NO_SPACE";
            case BiometricPrompt.ERROR_SECURITY_UPDATE_REQUIRED:
                return "ERROR_SECURITY_UPDATE_REQUIRED";
            case BiometricPrompt.ERROR_TIMEOUT:
                return "ERROR_TIMEOUT";
            case BiometricPrompt.ERROR_UNABLE_TO_PROCESS:
                return "ERROR_UNABLE_TO_PROCESS";
            case BiometricPrompt.ERROR_USER_CANCELED:
                return "ERROR_USER_CANCELED";
            case BiometricPrompt.ERROR_VENDOR:
                return "ERROR_VENDOR";
            default:
                return "ERROR_UNKNOWN";
        }
    }

    private void reject(PluginCall call, Exception e) {
        call.reject(e.getMessage(), e.getClass().getName(), e);
    }

    private void rejectWithCancel(PluginCall call) {
        call.reject("CANCEL", "CANCEL");
    }
}

File: packages/authgear-capacitor/ios/Plugin/AuthgearPlugin.swift
import Foundation
import LocalAuthentication
import Capacitor

@objc(AuthgearPlugin)
public class AuthgearPlugin: CAPPlugin {
    private let impl = AuthgearPluginImpl()

    @objc func storageGetItem(_ call: CAPPluginCall) {
        let key = call.getString("key")!
        do {
            let value = try self.impl.storageGetItem(key: key)
            call.resolve([
                "value": value,
            ])
        } catch {
            let nsError = error as NSError
            if nsError.domain == NSOSStatusErrorDomain && nsError.code == errSecItemNotFound {
                call.resolve([
                    "value": nil
                ])
            } else {
                error.reject(call)
            }
        }
    }

    @objc func storageSetItem(_ call: CAPPluginCall) {
        let key = call.getString("key")!
        let value = call.getString("value")!
        do {
            try self.impl.storageSetItem(key: key, value: value)
            call.resolve()
        } catch {
            error.reject(call)
        }
    }

    @objc func storageDeleteItem(_ call: CAPPluginCall) {
        let key = call.getString("key")!
        do {
            try self.impl.storageDeleteItem(key: key)
            call.resolve()
        } catch {
            error.reject(call)
        }
    }

    @objc func randomBytes(_ call: CAPPluginCall) {
        let length = call.getInt("length")!
        do {
            let bytes = try self.impl.randomBytes(length: length)
            call.resolve([
                "bytes": bytes,
            ])
        } catch {
            error.reject(call)
        }
    }

    @objc func sha256String(_ call: CAPPluginCall) {
        let input = call.getString("input")!
        do {
            let bytes = try self.impl.sha256String(input: input)
            call.resolve([
                "bytes": bytes,
            ])
        } catch {
            error.reject(call)
        }
    }

    @objc func getDeviceInfo(_ call: CAPPluginCall) {
        let deviceInfo = self.impl.getDeviceInfo()
        call.resolve([
            "deviceInfo": deviceInfo
        ])
    }

    @objc func generateUUID(_ call: CAPPluginCall) {
        let uuid = self.impl.generateUUID()
        call.resolve([
            "uuid": uuid
        ])
    }

    @objc func openAuthorizeURL(_ call: CAPPluginCall) {
        let url = URL(string: call.getString("url")!)!
        let callbackURL = URL(string: call.getString("callbackURL")!)!
        let prefersEphemeralWebBrowserSession = call.getBool("prefersEphemeralWebBrowserSession")!

        DispatchQueue.main.async {
            self.impl.openAuthorizeURL(window: (self.bridge?.webView?.window)!, url: url, callbackURL: callbackURL, prefersEphemeralWebBrowserSession:                prefersEphemeralWebBrowserSession) { (redirectURI, error) in
                if let error = error {
                    error.reject(call)
                }
                if let redirectURI = redirectURI {
                    call.resolve([
                        "redirectURI": redirectURI
                    ])
                }
            }
        }
    }

    @objc func openAuthorizeURLWithWebView(_ call: CAPPluginCall) {
        let url = URL(string: call.getString("url")!)!
        let redirectURI = URL(string: call.getString("redirectURI")!)!
        let modalPresentationStyle = call.getString("modalPresentationStyle")
        let navigationBarBackgroundColor = call.getString("navigationBarBackgroundColor")
        let navigationBarButtonTintColor = call.getString("navigationBarButtonTintColor")
        let isInspectable = call.getBool("iosIsInspectable")

        DispatchQueue.main.async {
            self.impl.openAuthorizeURLWithWebView(
                window: (self.bridge?.webView?.window)!,
                url: url,
                redirectURI: redirectURI,
                modalPresentationStyleString: modalPresentationStyle,
                navigationBarBackgroundColorString: navigationBarBackgroundColor,
                navigationBarButtonTintColorString: navigationBarButtonTintColor,
                isInspectable: isInspectable
            ) { (redirectURI, error) in
                if let error = error {
                    error.reject(call)
                }
                if let redirectURI = redirectURI {
                    call.resolve([
                        "redirectURI": redirectURI
                    ])
                }
            }
        }
    }

    @objc func openURL(_ call: CAPPluginCall) {
        let url = URL(string: call.getString("url")!)!

        DispatchQueue.main.async {
            self.impl.openURL(window: (self.bridge?.webView?.window)!, url: url) { (error) in
                if let error = error {
                    error.reject(call)
                } else {
                    call.resolve()
                }
            }
        }
    }

    @objc func checkBiometricSupported(_ call: CAPPluginCall) {
        DispatchQueue.main.async {
            do {
                try self.impl.checkBiometricSupported()
                call.resolve()
            } catch {
                error.reject(call)
            }
        }
    }

    @objc func createBiometricPrivateKey(_ call: CAPPluginCall) {
        let kid = call.getString("kid")!
        let payload = call.getObject("payload")!
        let ios = call.getObject("ios")!
        let constraint = ios["constraint"] as! String
        let localizedReason = ios["localizedReason"] as! String
        let tag = "com.authgear.keys.biometric.\(kid)"
        let policy = LAPolicy.deviceOwnerAuthenticationWithBiometrics

        DispatchQueue.main.async {
            self.impl.createBiometricPrivateKey(
                policy: policy,
                localizedReason: localizedReason,
                constraint: constraint,
                kid: kid,
                tag: tag,
                payload: payload
            ) { (jwt, error) in
                if let error = error {
                    error.reject(call)
                }
                if let jwt = jwt {
                    call.resolve([
                        "jwt": jwt
                    ])
                }
            }
        }
    }

    @objc func signWithBiometricPrivateKey(_ call: CAPPluginCall) {
        let kid = call.getString("kid")!
        let payload = call.getObject("payload")!
        let ios = call.getObject("ios")!
        let policyString = ios["policy"] as! String
        let localizedReason = ios["localizedReason"] as! String
        let tag = "com.authgear.keys.biometric.\(kid)"

        DispatchQueue.main.async {
            self.impl.signWithBiometricPrivateKey(
                policyString: policyString,
                localizedReason: localizedReason,
                kid: kid,
                tag: tag,
                payload: payload
            ) { (jwt, error) in
                if let error = error {
                    error.reject(call)
                }
                if let jwt = jwt {
                    call.resolve([
                        "jwt": jwt
                    ])
                }
            }
        }
    }

    @objc func removeBiometricPrivateKey(_ call: CAPPluginCall) {
        let kid = call.getString("kid")!
        let tag = "com.authgear.keys.biometric.\(kid)"

        DispatchQueue.main.async {
            do {
                try self.impl.removeBiometricPrivateKey(tag: tag)
                call.resolve()
            } catch {
                error.reject(call)
            }
        }
    }

    @objc func createDPoPPrivateKey(_ call: CAPPluginCall) {
        let kid = call.getString("kid")!

        DispatchQueue.main.async {
            do {
                try self.impl.createDPoPPrivateKey(kid: kid)
                call.resolve()
            } catch {
                error.reject(call)
            }
        }
    }

    @objc func signWithDPoPPrivateKey(_ call: CAPPluginCall) {
        let kid = call.getString("kid")!
        let payload = call.getObject("payload")!

        DispatchQueue.main.async {
            do {
                let jwt = try self.impl.signWithDPoPPrivateKey(kid: kid, payload: payload)
                call.resolve(["jwt": jwt])
            } catch {
                error.reject(call)
            }
        }
    }

    @objc func checkDPoPPrivateKey(_ call: CAPPluginCall) {
        let kid = call.getString("kid")!

        DispatchQueue.main.async {
            let ok = self.impl.checkDPoPPrivateKey(kid: kid)
            call.resolve(["ok": ok])
        }
    }

    @objc func computeDPoPJKT(_ call: CAPPluginCall) {
        let kid = call.getString("kid")!

        DispatchQueue.main.async {
            do {
                let jkt = try self.impl.computeDPoPJKT(kid: kid)
                call.resolve(["jkt": jkt])
            } catch {
                error.reject(call)
            }
        }
    }
}
